### PR TITLE
SMT12 offline migrations

### PR DESCRIPTION
**Pull request by ikapelyukhin**

### CODE REVIEW

**Comment by ikapelyukhin** [comment]
Something's wrong when getting list of possible migrations with Yast:

image


```
[Ruby] ui/wizard_client.rb:55 Caught error: NoMethodError: "undefined method `friendly_name' for nil:NilClass",
 ["/usr/share/YaST2/lib/registration/ui/migration_selection_dialog.rb:185:in `base_product_text_for'",
 "/usr/share/YaST2/lib/registration/ui/migration_selection_dialog.rb:160:in `block in migration_items'",
 "/usr/share/YaST2/lib/registration/ui/migration_selection_dialog.rb:159:in `map'",
 "/usr/share/YaST2/lib/registration/ui/migration_selection_dialog.rb:159:in `with_index'",
 "/usr/share/YaST2/lib/registration/ui/migration_selection_dialog.rb:159:in `migration_items'",
 "/usr/share/YaST2/lib/registration/ui/migration_selection_dialog.rb:153:in `migration_selection_widget'",
 "/usr/share/YaST2/lib/registration/ui/migration_selection_dialog.rb:132:in `dialog_content'",
 "/usr/share/YaST2/lib/registration/ui/migration_selection_dialog.rb:64:in `run'",
 "/usr/share/YaST2/lib/registration/ui/migration_repos_workflow.rb:383:in `select_migration_products'",
 "/usr/share/YaST2/lib/registration/ui/migration_repos_workflow.rb:84:in `block in run_sequence'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/builtins.rb:546:in `eval'",
 "/update/000/inst-sys/usr/share/YaST2/modules/Sequencer.rb:261:in `WS_run'",
 "/update/000/inst-sys/usr/share/YaST2/modules/Sequencer.rb:333:in `block in Run'",
 "/update/000/inst-sys/usr/share/YaST2/modules/Sequencer.rb:325:in `loop'",
 "/update/000/inst-sys/usr/share/YaST2/modules/Sequencer.rb:325:in `Run'",
 "/usr/share/YaST2/lib/registration/ui/migration_repos_workflow.rb:93:in `run_sequence'",
 "/usr/share/YaST2/lib/registration/ui/wizard_client.rb:51:in `run'",
 "/usr/share/YaST2/lib/registration/ui/wizard_client.rb:41:in `main'",
 "/update/000/inst-sys/usr/share/YaST2/clients/migration_repos.rb:17:in `<top (required)>'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `eval'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `run_client'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin_wrapper'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:195:in `CallFunction'",
 "/usr/share/YaST2/lib/registration/clients/inst_migration_repos.rb:35:in `main'",
 "/update/000/inst-sys/usr/share/YaST2/clients/inst_migration_repos.rb:17:in `<top (required)>'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `eval'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `run_client'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin_wrapper'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:195:in `CallFunction'",
 "/usr/share/YaST2/lib/registration/ui/offline_migration_workflow.rb:136:in `migration_repos'",
 "/usr/share/YaST2/lib/registration/ui/offline_migration_workflow.rb:57:in `main'",
 "/update/000/inst-sys/usr/share/YaST2/clients/inst_offline_migration.rb:17:in `<top (required)>'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `eval'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `run_client'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin_wrapper'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:195:in `CallFunction'",
 "/update/000/inst-sys/usr/share/YaST2/modules/ProductControl.rb:1355:in `RunFrom'",
 "/update/000/inst-sys/usr/share/YaST2/clients/inst_update_partition.rb:66:in `main'",
 "/update/000/inst-sys/usr/share/YaST2/clients/inst_update_partition.rb:105:in `<top (required)>'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `eval'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `run_client'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin_wrapper'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:195:in `CallFunction'",
 "/update/000/inst-sys/usr/share/YaST2/modules/ProductControl.rb:1355:in `RunFrom'",
 "/update/000/inst-sys/usr/share/YaST2/clients/inst_update_partition.rb:66:in `main'",
 "/update/000/inst-sys/usr/share/YaST2/clients/inst_update_partition.rb:105:in `<top (required)>'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `eval'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `run_client'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin_wrapper'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:195:in `CallFunction'",
 "/update/000/inst-sys/usr/share/YaST2/modules/ProductControl.rb:1355:in `RunFrom'",
 "/update/000/inst-sys/usr/share/YaST2/clients/inst_update_partition.rb:66:in `main'",
 "/update/000/inst-sys/usr/share/YaST2/clients/inst_update_partition.rb:105:in `<top (required)>'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `eval'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `run_client'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin_wrapper'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:195:in `CallFunction'",
 "/update/000/inst-sys/usr/share/YaST2/modules/ProductControl.rb:1355:in `RunFrom'",
 "/update/000/inst-sys/usr/share/YaST2/modules/ProductControl.rb:1542:in `Run'",
 "/usr/share/YaST2/lib/installation/clients/inst_worker_initial.rb:106:in `main'",
 "/update/000/inst-sys/usr/share/YaST2/clients/inst_worker_initial.rb:2:in `<top (required)>'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `eval'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `run_client'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin_wrapper'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:195:in `CallFunction'",
 "/usr/share/YaST2/lib/installation/clients/installation.rb:81:in `main'",
 "/update/000/inst-sys/usr/share/YaST2/clients/installation.rb:2:in `<top (required)>'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `eval'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:313:in `run_client'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:206:in `call_builtin_wrapper'",
 "/usr/lib64/ruby/vendor_ruby/2.5.0/yast/wfm.rb:195:in `CallFunction'",
 "/usr/lib/YaST2/bin/y2start:58:in `<main>'"]
[Ruby] ui/wizard_client.rb:57 Internal error: undefined method `friendly_name' for nil:NilClass
```

**Comment by ikapelyukhin** [comment]
TODO: `upgrade` endpoint needs to be updated to allow for additional products to be activated during the upgrade.

File: www/perl-lib/SMT/Rest/SCCAPIv4.pm [hunk 1]
```diff
@@ -37,6 +37,7 @@ sub systems_handler()
         elsif ( $self->request()->method() =~ /^POST$/i )
         {
             if     ( $path =~ /^systems\/products\/migrations\/?$/ ) { return $self->product_migration_targets(); }
+            elsif  ( $path =~ /^systems\/products\/offline_migrations\/?$/ ) { return $self->product_migration_targets(1); }
```
Comment (felixsch) [inline www/perl-lib/SMT/Rest/SCCAPIv4.pm]: nitpicking: why not define constants,  which reflect what product_migration is mode is running. I think boolean/int should be avoided when defining behavior. 
```
$self->product_migration_targets(MIGRATION_OFFLINE);
$self->product_migration_targets(MIGRATION_ONLINE);
```
just a thought!

Comment (ikapelyukhin) [reply www/perl-lib/SMT/Rest/SCCAPIv4.pm]: Yeah, makes sense. Should be more appropriate than passing an `online`/`offline` string for the migration kind, too.

Comment (ikapelyukhin) [reply www/perl-lib/SMT/Rest/SCCAPIv4.pm]: Fixed.

**Review by felixsch — APPROVED** [review]
lgtm!

**Pull request merged**